### PR TITLE
Check for permissions before creating or editing rules on backend

### PR DESCRIPTION
**Pull request by rhystmills**
## What does this change?

1. Check for create and edit permissions on the backend before accepting a users request.
2. Only allow users to request rule creation or edit changes on the frontend if they have correct permissions.

Kapture 2023-05-15 at 10 56 40

## How to test

1. Run Typerighter locally according to the instructions in the readme.
2. Visit [manager.typerighter.local.dev-gutools.co.uk](https://manager.typerighter.local.dev-gutools.co.uk/)
3. Make sure you have 'Manage rules' Typerighter permissions [on permissions CODE](https://permissions.code.dev-gutools.co.uk/admin/user/111318226344948727644). Wait a minute for them to apply. Can you create and edit rules in the Typerighter manager locally?
4. Remove the permissions and wait a minute or so for the changes to persist. Can you no longer create and edit rules locally in Typerighter manager?

### CODE REVIEW

File: apps/rule-manager/client/src/ts/components/RuleForm.tsx [hunk 1]
```diff
@@ -45,6 +45,7 @@ export const RuleForm = ({onRuleUpdate, ruleData, setRuleData, createRuleFormOpe
     }) => {
     const [showErrors, setShowErrors] = useState(false);
     const [errors, setErrors] = useState<FormError[]>([]);
+    const pageData = useContext(PageContext)
```
Comment (jonathonherbert) [inline apps/rule-manager/client/src/ts/components/RuleForm.tsx:48]: Non-blocking: just thinking that we do this in two places – we could make this slightly more concise by writing a `useCreateEditPermissions` hook that looks a bit like:

```ts
const useCreateEditPermissions = () => {
  const pageData = useContext(PageContext);
  // Do not recalculate permissions if the permissions list has not changed
  return useMemo(() => hasCreateEditPermissions(pageData.permissions), [pageData.permissions]);
}
```

(The memoisation optimisation is not super important, but arguably a good habit when calling pure fns from React components.)

File: apps/rule-manager/client/src/ts/components/RulesTable.tsx [hunk 2]
```diff
@@ -104,7 +126,8 @@ const RulesTable = () => {
                 }
             })
     }
-    const columns = createColumns(openEditRulePanel);
+    const pageData = useContext(PageContext);
+    const columns = createColumns(openEditRulePanel, pageData ? hasCreateEditPermissions(pageData.permissions) : false);
```
Comment (jonathonherbert) [inline apps/rule-manager/client/src/ts/components/RulesTable.tsx:130]: Non-blocking: you could move the permissions hooks within `createColumns`, and save passing an argument here.

**Review by jonathonherbert — APPROVED** [review]
Tested client and server behaviour, works as expected. Reckon tooltips are a nice solution here, really clear! 🎉

File: apps/rule-manager/client/src/ts/components/RulesTable.tsx (hunk 2 above)
Comment (rhystmills) [reply apps/rule-manager/client/src/ts/components/RulesTable.tsx]: Now done!

File: apps/rule-manager/client/src/ts/components/RuleForm.tsx (hunk 1 above)
Comment (rhystmills) [reply apps/rule-manager/client/src/ts/components/RuleForm.tsx]: Now done.

**Review by jonathonherbert — APPROVED** [review]

**Pull request merged**